topology/controllers.go

```diff
@@ -0,0 +1,111 @@
+package topology
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"text/template"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/incident-commander/api"
+	babel "github.com/jvatic/goja-babel"
+	"github.com/labstack/echo/v4"
+)
+
+var jsComponentTpl *template.Template
+
+func init() {
+	tpl, err := template.New("registry").Parse(jsComponentRegistryTpl)
+	if err != nil {
+		logger.Fatalf("error parsing template 'jsComponentRegistryTpl'. %v", err)
+	}
+
+	jsComponentTpl = tpl
+}
+
+type component struct {
+	Name string
+	JS   string
+}
+
+// GetCustomRenderer returns an application/javascript HTTP response
+// with custom components and a registry.
+// This registry needs to be used to select custom components
+// for rendering of properties and cards.
+func GetCustomRenderer(ctx echo.Context) error {
+	// 1. Read the template of the topology
+	sysTplID := ctx.QueryParams().Get("id")
+	results, err := QueryRenderComponents(ctx.Request().Context(), sysTplID)
+	if err != nil {
+		return errorResonse(ctx, err, http.StatusBadRequest)
+	}
+
+	// 2. Create a registry of all the components
+	var components = make(map[string]component)
+	for _, r := range results {
+		if err := compileComponents(components, r.Components, false); err != nil {
+			return errorResonse(ctx, err, http.StatusInternalServerError)
```

Comment:
You can use
 ```go
		return c.JSON(http.StatusInternalServerError, api.HTTPErrorMessage{
			Error:   err.Error(),
			Message: "Error message for user to see",
		})
```

or wrap this in your errorResponse function, but prefer using api.HTTPErrorMessage